working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+is determined by the static type of the receiver. Similarly, like extension
+methods, members of an extension type are resolved statically, i.e., if
+`tiny.leaves` is an invocation of an extension type getter `leaves`, then
+the declaration named `leaves` whose body is executed is determined at
+compile-time. There is no support for late binding of an extension method,
+and hence there is no notion of overriding. In return for this lack of
+expressive power, we get improved performance.
+
+Here is another example. It illustrates the fact that an extension type `E`
+with on-type `T` introduces a type `E` which is a supertype of `T`. This
+makes it possible to assign an expression of type `T` to a variable of type
+`E`. This corresponds to "entering" the extension type (accepting the
+specific discipline associated with `E`). Conversely, a cast from `E` to
+`T` is a downcast, and hence it must be written explicitly. This cast
+corresponds to "exiting" the extension type (allowing for violations of the
+discipline associated with `E`), and the fact that the cast must be written
+explicitly helps developers maintaining the discipline as intended, rather
+than dropping out of it by accident, silently.
+
+```dart
+extension type ListSize<X> on List<X> {
+  int get size => length;
+  X front() => this[0];
+}
+
+void main() {
+  ListSize<String> xs = <String>['Hello']; // OK, upcast.
+  print(xs); // OK, `toString()` available on Object?.
+  print("Size: ${xs.size}. Front: ${xs.front()}"); // OK.
+  xs[0]; // Error, `operator []` not a member of `ListSize`.
+
+  List<ListSize<String>> ys = [xs]; // OK.
+  List<List<String>> ys2 = ys; // Error, downcast.
+  ListSize<ListSize<Object>> ys3 = ys; // OK.
+  ys[0].front(); // OK.
+  ys3.front().front(); // OK.
+  ys as List<List<String>>; // `ys` is promoted, succeeds at run time.
+}
+```
+
+
+## Syntax
+
+The rule for `<extensionDeclaration>` in the grammar is replaced by the
+following:
+
+```ebnf
+<extensionDeclaration> ::=
+  'protected'? 'extension' ('type'? <identifier> <typeParameters>?)?
+      <extensionShowHidePart> 'on' <type> <interfaces>? '{'
+    (<metadata> <extensionMemberDefinition>)* 
+  '}'
+
+<extensionShowHidePart> ::=
+  <extensionShowClause>? <extensionHideClause>?
+
+<extensionShowClause> ::= 'show' <extensionShowHideList>
+
+<extensionHideClause> ::= 'hide' <extensionShowHideList>
+
+<extensionShowHideList> ::= 
+  <extensionShowHideElement> (',' <extensionShowHideElement>)*
+
+<extensionShowHideElement> ::=
+  <type> | <identifier> | 'operator' <operator>
+```
+
+*Note that `<type>` derives `<typeIdentifier>`, which makes `<identifier>`
+nearly redundant. However, `<identifier>` is still needed because it
+includes some strings that cannot be the name of a type, e.g., the built-in
+identifiers.*
+
+
+## Static Analysis
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
```

Comment:
I'm a bit confused here.  Are you proposing to change normal extensions to define types now? That feels very surprising.

Reply:
As noted in discussion, I think my concerns boil down to the following.

First, it's not clear to me that this will work out.  It's not clear to me that it won't, but we need to be sure that the semantics we want are the ones we get already from extension methods, and that we won't break existing code by doing this.

Second, it's not clear to me that we want this.  A whole bunch of people wrote extension methods without expecting to be introducing a new type.  How will this affect existing code (and future uses of extension methods).  Maybe it's ok?  You can sort of think of it as a more convenient way of applying the explicit extension syntax?  But it needs some substantial discussion, I think.

Reply:
I'm fairly sure it will work. I considered it while designing extension declarations. That's also why explicit extension invocation looks like a constructor invocation, I wanted `Ext(o)` to be a valid expression of the extension type if we introduce extension types with constructors. Doing `Ext(o).method` and `(o as Ext).method` will mean the same thing.

Whether we want it ... 
It's true that the extensions were not designed to be used as types, and I'm slightly worried about unintended uses being potentially a longer term support issue (like using classes as mixins even if they weren't intended for it).
The biggest argument against is that most extensions don't work particularly *well* as types. They just add individual methods. I'd want to do a `show *` by default on those types so that the extension type actually works just like calling directly on the object, just with only extensions from  that extension. Hiding the object's own methods is not going to be useful.
The names also don't sound like type names.

Reply:
We could use the keyword `type` to bundle the two things: (1) use as a type, and (2) enable implicit invocations. We would then have the following combinations:

- Without `type`: "You cannot use it as a type, but you can invoke members implicitly". This is the traditional extension declaration.
- With `type`: "You can use it as a type, but you cannot invoke members implicitly".